### PR TITLE
Remove navigator.deviceMemory

### PR DESCRIPTION
**Pull request by bershanskiy**
Fixes [#1157](https://github.com/brave/brave-browser/issues/1157).

# Tests

I manually tested that `navigator.deviceMemory` is missing (undefined) in all contexts, including- first and third-party frames and workers.

The above issue also has patches for Chromium unit tests, but they were excluded for brevity.

When I follow Tests wiki article and run unit tests, they get skipped because there are "no test result for" any of the tests.

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [x] Linux
  - [ ] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

**Comment by snyderp** [comment]
One note on this though, this implementation disables permanently; it doesn't put it behind shields.  Thats a-ok with me, but @bsclifton is that a concern product side?

**Comment by bershanskiy** [comment]
> One note on this though, this implementation disables permanently; it doesn't put it behind shields. Thats a-ok with me...

I personally would prefer if all the fingerprinting features were behind individual flags (with global on/off toggle that allows everything and individual flags for every feature when shields are up + some sensible defaults). Also, it would be nice to see when a particular fingerprintable feature is accessed for both troubleshooting and to assess site's intentions. But that would require significantly more work.

**Comment by jumde** [comment]
@bershanskiy @bsclifton - Removing the api completely might lead to webcompat issues, we should probably set it to return 1 always. See: https://github.com/brave/brave-browser/issues/1157#issuecomment-434860218

**Comment by snyderp** [comment]
@jumde i think its very unlikely to cause web comapt issues, since no one other than chrome supports

**Comment by snyderp** [comment]
I have a mild preference for removing over fixed value, since for the minority of sites that query the value, they may make decisions based on 1GB (that might cause problems on devices with less mem), while removing the end point would push people onto the Safari / FF / more conservative path.

In other words, unless removing the feature is likely to cause webcompat problems (I dont think the case here) I think we should remove vs fixed reply, to (slightly) push against further adoption of the harmful feature

**Comment by jumde** [comment]
That sounds good to me. cc: @diracdeltas @riastradh-brave - do you have any concerns about removing this API?

**Comment by riastradh-brave** [comment]
My only concern was that someone else might object to removing it!  (At the time I wasn't sure how often it would be used and was waiting for Pete to provide empirical data from crawls.)  Can't wait to be rid of more fingerprinting vectors.

**Comment by snyderp** [comment]
+1 for axing

**Comment by bsclifton** [comment]
cc: @bridiver for patch review

File: patches/third_party-blink-renderer-core-frame-navigator.h.patch [hunk 1]
```diff
@@ -0,0 +1,20 @@
+diff --git a/third_party/blink/renderer/core/frame/navigator.h b/third_party/blink/renderer/core/frame/navigator.h
+index 4572df017cf4..377827ad9a3c 100644
+--- a/third_party/blink/renderer/core/frame/navigator.h
++++ b/third_party/blink/renderer/core/frame/navigator.h
+@@ -24,7 +24,6 @@
+ #include "third_party/blink/renderer/core/core_export.h"
+ #include "third_party/blink/renderer/core/execution_context/context_lifecycle_observer.h"
+ #include "third_party/blink/renderer/core/frame/navigator_concurrent_hardware.h"
+-#include "third_party/blink/renderer/core/frame/navigator_device_memory.h"
```
Comment (bridiver) [inline patches/third_party-blink-renderer-core-frame-navigator.h.patch]: we are discussing exactly what we want to do here (returning a fixed value vs removing the api, but either way removing this header is not necessary and I'm not even sure if removing https://github.com/brave/brave-core/pull/3669/files#diff-231a100e06059effa4a4b129d96e35fcR17 is necessary. Is https://github.com/brave/brave-core/pull/3669/files#diff-04ad001bca876052b0b18cc9dfa3df9dR9 alone not sufficient?

File: patches/third_party-blink-renderer-core-frame-navigator.idl.patch [hunk 2]
```diff
@@ -0,0 +1,12 @@
+diff --git a/third_party/blink/renderer/core/frame/navigator.idl b/third_party/blink/renderer/core/frame/navigator.idl
+index f13febb148e9..b8e0bbc2b66e 100644
+--- a/third_party/blink/renderer/core/frame/navigator.idl
++++ b/third_party/blink/renderer/core/frame/navigator.idl
+@@ -36,7 +36,6 @@
+ 
+ Navigator includes NavigatorConcurrentHardware;
+ Navigator includes NavigatorCookies;
+-Navigator includes NavigatorDeviceMemory;
```
Comment (bridiver) [inline patches/third_party-blink-renderer-core-frame-navigator.idl.patch]: this appears to be sufficient, but can you comment it out instead `// Navigator includes NavigatorDeviceMemory;`?

**Comment by bershanskiy** [comment]
FYI, I incorporated the review feedback and rebased the PR on current master to simplify testing.

File: patches/third_party-blink-renderer-core-frame-navigator.h.patch (hunk 1 above)
Comment (bershanskiy) [reply patches/third_party-blink-renderer-core-frame-navigator.h.patch]: You are right, this patch is unnecessary, I removed it.

File: patches/third_party-blink-renderer-core-frame-navigator.idl.patch (hunk 2 above)
Comment (bershanskiy) [reply patches/third_party-blink-renderer-core-frame-navigator.idl.patch]: Done.

**Comment by bershanskiy** [comment]
As is, this simple patch compiles and removes `navigator.deviceMemory`, but does not have any tests. Should I add tests? Is there final verdict on "remove vs. present fixed value"?

I personally would prefer if APIs like this functioned like a trip wire for fingerprinting code. That is the API is "available", but when FP protection is on, it serves a fixed value and logs all queries somewhere (internally and to display in "Shields"). Then user can either allow a particular API (if something appears broken) and see how much finger-printable information is queried. If lot of attributes queried without apparent reason, then this is probably fingerprinting script. Of course, this would require much more granular `BraveContentSettingsObserver::AllowFingerprinting` than the current one, more logging and more UI work.

**Comment by bsclifton** [comment]
@bershanskiy tests would be great, but understandably they may be complex to add. I think we can merge this without tests, but if you wanted to look at writing a browser test, you should be able to create a simple HTML/JavaScript proof of concept that the test would load. Then I can help you find a similar test

**Comment by bershanskiy** [comment]
@bsclifton 

> Checking with the team on remove vs fake value

Great, [my earlier comment](https://github.com/brave/brave-core/pull/3669#issuecomment-572812224) still stands.

> I think we can merge this without tests

Then let's merge this as is, if it's OK.

About tests:
I'll add tests later, when I figure out more about them.
The tests indeed would be trivial, like `navigator.deviceMemory === undefined`, but the hardest part is integrating them into the Brave test suite. Also, do you patch Chromium's own tests? E.g., there are about [20 files called `*-expected.txt`](https://github.com/chromium/chromium/search?p=1&q=deviceMemory&unscoped_q=deviceMemory) that check for `deviceMemory` (among other things) and tests should fail if they are not patched. It appears as if Brave does not have such patches, but I just wanted to make sure.

**Comment by bridiver** [comment]
@bershanskiy we generally don't patch chromium tests unless we're planning to upstream a feature, but you can use them as a basis for your own test and there are existing tests for deviceMemory that could be copied

**Comment by bsclifton** [comment]
@bershanskiy we're good to go with this approach (conferred w/ team) 👍 

The tests you've found are part of the base Chromium code and we're not running them. Although, it would be good to try to do so during our continuous integration process (but that's another issue)

**Pull request merged**